all_gather_bucketing fx pass

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #157396


Porting passes to bucket all_gathers

The main logic of the pass is done via 
1. Searching for all all_gathers from the buckets


Copying tests from @wconstab PR to test compatibility with reordering.
Test checks only compatibility, as because of (3) the joint all_gather will be scheduled already as early as possible and no space for reordering.

Pass changes:
Using mutation ops to match performance of fsdp, in future the perfect scenario will be to have only functional graph, that inductor does all memory optimizations on its own without mutable ops.


Inductor changes:
Adding foreach_copy_ lowering


cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

IvanKobzarev committed:
all_gather_bucketing fx pass

[ghstack-poisoned]

bdhirsh commented on torch/_inductor/fx_passes/fsdp.py:
```diff
@@ -0,0 +1,347 @@
+import logging
```
hmm if we want to land this as an initial port just so we have something to start iterating on i'm ok with that, but my vote would be to not hardcode FSDP in the file / pass names. Since theoretically this should be a generic pass for doing bucketing (and when we want to later do bucketing for allreduces, we should hopefully be able to re-use a big chunk of the code).

What do folks think of renaming this to `bucketing.py` or something? (and calling this a generic allgather_bucketing pass)

bdhirsh commented on torch/_inductor/fx_passes/fsdp.py:
```diff
@@ -0,0 +1,347 @@
+
+    # NOTE: this overwrites the existing post_grad_custom_post_pass
+    torch._inductor.config.post_grad_custom_post_pass = (
+        apply_simplefsdp_bucketing_passes
```
out of date?

IvanKobzarev replied:
Yeah, removed it.

IvanKobzarev committed:
Update on "all_gather_bucketing fx pass"



Porting passes to bucket all_gathers

The main logic of the pass is done via 
1. Searching for all all_gathers from the buckets
2. Inserting the trace of the function instead of all_gathers:
```
def all_gather_merge_fn_to_trace(  # type: ignore[no-untyped-def]
    ag_inputs, group_size: int, group_name: str, dtype: torch.dtype
) -> list[torch.Tensor]:
    inp_sizes = [ag_in.shape for ag_in in ag_inputs]
    inp_split_sizes = [ag_in.numel() for ag_in in ag_inputs]
    ag_input_numel = sum(inp_split_sizes)
    device = ag_inputs[0].device
    rank = device.index
    new_ag_out = torch.empty(ag_input_numel * group_size, dtype=dtype, device=device)
    foreach_copy_dsts = torch.split(new_ag_out, inp_split_sizes)
    ag_inputs_reshaped = [ag_in.reshape(-1) for ag_in in ag_inputs]
    torch._foreach_copy_(foreach_copy_dsts, ag_inputs_reshaped)
    new_ag_in = new_ag_out.narrow(0, ag_input_numel * rank, ag_input_numel)
    wait_tensor = torch.ops.c10d_functional.wait_tensor(
        torch.ops._c10d_functional.all_gather_into_tensor_out.default(
            new_ag_in, group_size, group_name, out=new_ag_out
        )
    )
    new_ag_out_reshaped = wait_tensor.reshape(group_size, -1)
    outs = torch.split_with_sizes(
        new_ag_out_reshaped,
        inp_split_sizes,
        dim=1,
    )
    outs_reshaped = [out.reshape(shape) for out, shape in zip(outs, inp_sizes)]
    return outs_reshaped
```
Having special function that we trace allows to simplify the experimentation on changing what we are writing.
3. This graph will be inserted as early as possible - just after we found the last input to bucketing all_gathers

Copying tests from wconstab PR to test compatibility with reordering.
Test checks only compatibility, as because of (3) the joint all_gather will be scheduled already as early as possible and no space for reordering.

Will try to  come up to test actual reordering.


Pass changes:
Using mutation ops to match performance of fsdp, in future the perfect scenario will be to have only functional graph, that inductor does all memory optimizations on its own without mutable ops.


Inductor changes:
Adding foreach_copy_ lowering


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "all_gather_bucketing fx pass"



Porting passes to bucket all_gathers

The main logic of the pass is done via 
1. Searching for all all_gathers from the buckets
2. Inserting the trace of the function instead of all_gathers:
```
def all_gather_merge_fn_to_trace(  # type: ignore[no-untyped-def]
    ag_inputs, group_size: int, group_name: str, dtype: torch.dtype
) -> list[torch.Tensor]:
    inp_sizes = [ag_in.shape for ag_in in ag_inputs]
    inp_split_sizes = [ag_in.numel() for ag_in in ag_inputs]
    ag_input_numel = sum(inp_split_sizes)
    device = ag_inputs[0].device
    rank = device.index
    new_ag_out = torch.empty(ag_input_numel * group_size, dtype=dtype, device=device)
    foreach_copy_dsts = torch.split(new_ag_out, inp_split_sizes)
    ag_inputs_reshaped = [ag_in.reshape(-1) for ag_in in ag_inputs]
    torch._foreach_copy_(foreach_copy_dsts, ag_inputs_reshaped)
    new_ag_in = new_ag_out.narrow(0, ag_input_numel * rank, ag_input_numel)
    wait_tensor = torch.ops.c10d_functional.wait_tensor(
        torch.ops._c10d_functional.all_gather_into_tensor_out.default(
            new_ag_in, group_size, group_name, out=new_ag_out
        )
    )
    new_ag_out_reshaped = wait_tensor.reshape(group_size, -1)
    outs = torch.split_with_sizes(
        new_ag_out_reshaped,
        inp_split_sizes,
        dim=1,
    )
    outs_reshaped = [out.reshape(shape) for out, shape in zip(outs, inp_sizes)]
    return outs_reshaped
```
Having special function that we trace allows to simplify the experimentation on changing what we are writing.
3. This graph will be inserted as early as possible - just after we found the last input to bucketing all_gathers

Copying tests from wconstab PR to test compatibility with reordering.
Test checks only compatibility, as because of (3) the joint all_gather will be scheduled already as early as possible and no space for reordering.

Will try to  come up to test actual reordering.


Pass changes:
Using mutation ops to match performance of fsdp, in future the perfect scenario will be to have only functional graph, that inductor does all memory optimizations on its own without mutable ops.


Inductor changes:
Adding foreach_copy_ lowering


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "all_gather_bucketing fx pass"



Porting passes to bucket all_gathers

The main logic of the pass is done via 
1. Searching for all all_gathers from the buckets


Copying tests from wconstab PR to test compatibility with reordering.
Test checks only compatibility, as because of (3) the joint all_gather will be scheduled already as early as possible and no space for reordering.

Pass changes:
Using mutation ops to match performance of fsdp, in future the perfect scenario will be to have only functional graph, that inductor does all memory optimizations on its own without mutable ops.


Inductor changes:
Adding foreach_copy_ lowering


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

wconstab commented on torch/_inductor/fx_passes/bucketing.py:
```diff
@@ -0,0 +1,431 @@
+            assert (
+                ag_node in ag_node_to_wait_node
+                and ag_node.args[1] == group_size
+                and ag_node.args[2] == group_name
```
this assumption is unfortunately too narrow.  In the FSDP + TP case, there will be all_fathers on multiple process groups, and this assert fails.

I wonder if anybody has considered TP all_gathers yet in bucketing pass design?
cc @ruisizhang123

wconstab replied:
There might be a simple fix here.
What if instead of asserting on 'same group', we just tweak the logic a bit:
- one "bucket" can now contain allgathers from any number of processgroups
- when it comes time to issue the 'bucketed allgather', instead of issuing one allgather per bucket, we issue one allgather per process-group per bucket
- bucketing assignment logic could be left alone for now

In the future, bucketing assignment logic could be improved, for example if we have the pattern of 
(TP-allgather, FSDP-allgather) for each param, it probably makes sense to either put both of those into a bucket or neither of them.  But for now we can let them be separated, i think.

ruisizhang123 replied:
1. In inductor, we look into the metadata trace of each IR node. If it contains `Reparameterization`, which is why SimpleFSDP ensures the graph traceable, we consdier the node as all-gather node. 

> considered TP all_gathers yet in bucketing pass design

2. You mean bucket TP all-gathers?  I didn't try, but seems doable to me lolll.

wconstab replied:
when we combine TP + FSDP and perform bucketing, it seems pointless / impossible to perform any bucketing on FSDP collectives if we do not also perform bucketing on TP collectives.

or am i missing something? did we handle TP a different way in the simple-fsdp experiments?

ruisizhang123 replied:
Yes, it triggers 2 separate AGs. Not sure there are any difference in doing this in FX-level versus Inductor-level. In inductor, we just pick out the FSDP AGs, and bucket them. It prefetches the one-dim parameters ahead of time. The rest TP AGs gathers things in their original order.  Not sure there will be any blockers for doing this? 

This is what the profile trace looks like (The upper one is FSDP, the lower one is TP)

<img width="1185" alt="Screenshot 2025-07-02 at 4 17 48 PM" src="https://github.com/user-attachments/assets/672fbab8-b6cb-4f47-bb0f-b48a8386d53a" />

wconstab replied:
sorry, i have just been saying nonsense. 

there shouldn't even be allgathers for TP, should there?

I think what was happening to me is the autoparallel solver was picking to do FSDP on the TP dim, and so it was really like fsdp(fsdp(param)).  If there is TP, we should be doing the sharded matmul and an allreduce..

ruisizhang123 replied:
yes, sorry for not familiar with tp comm details....

wconstab commented on torch/_inductor/fx_passes/bucketing.py:
```diff
@@ -0,0 +1,431 @@
+                    new_graph,
+                    torch.ops.fsdp.all_gather_copy_in.default,
+                    (
+                        param_all_gather_inputs_flattened,
```
hmm, has this code been tested? it seems wrong to me, and errors if i try to run it
```
  File "/data/users/whc/pytorch/torch/_inductor/fx_passes/bucketing.py", line 356, in merge_all_gather
    all_gather_copy_in = new_graph_call_function(
  File "/data/users/whc/pytorch/torch/_inductor/fx_passes/bucketing.py", line 150, in new_graph_call_function
    new_fake_tensor = target(*args_val, **kwargs_val)
  File "/data/users/whc/pytorch/torch/_ops.py", line 829, in __call__
    return self._op(*args, **kwargs)
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
RuntimeError: fsdp::all_gather_copy_in() expected at most 5 argument(s) but received 9 argument(s). Declaration: fsdp::all_gather_copy_in(Tensor[] all_gather_inputs, Tensor all_gather_output, SymInt[] inp_split_sizes, SymInt all_gather_input_numel, SymInt rank) -> (Tensor, Tensor)

```

indeed the op appears to accept only 5 arguments

IvanKobzarev replied:
Updated the PR to not use fsdp.all_gather_copy_in inlining the logic into fx graph.

IvanKobzarev committed:
Update on "all_gather_bucketing fx pass"



Porting passes to bucket all_gathers

The main logic of the pass is done via 
1. Searching for all all_gathers from the buckets


Copying tests from wconstab PR to test compatibility with reordering.
Test checks only compatibility, as because of (3) the joint all_gather will be scheduled already as early as possible and no space for reordering.

Pass changes:
Using mutation ops to match performance of fsdp, in future the perfect scenario will be to have only functional graph, that inductor does all memory optimizations on its own without mutable ops.


Inductor changes:
Adding foreach_copy_ lowering


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "all_gather_bucketing fx pass"



Porting passes to bucket all_gathers

The main logic of the pass is done via 
1. Searching for all all_gathers from the buckets


Copying tests from wconstab PR to test compatibility with reordering.
Test checks only compatibility, as because of (3) the joint all_gather will be scheduled already as early as possible and no space for reordering.

Pass changes:
Using mutation ops to match performance of fsdp, in future the perfect scenario will be to have only functional graph, that inductor does all memory optimizations on its own without mutable ops.


Inductor changes:
Adding foreach_copy_ lowering


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "all_gather_bucketing fx pass"



Porting passes to bucket all_gathers

The main logic of the pass is done via 
1. Searching for all all_gathers from the buckets


Copying tests from wconstab PR to test compatibility with reordering.
Test checks only compatibility, as because of (3) the joint all_gather will be scheduled already as early as possible and no space for reordering.

Pass changes:
Using mutation ops to match performance of fsdp, in future the perfect scenario will be to have only functional graph, that inductor does all memory optimizations on its own without mutable ops.


Inductor changes:
Adding foreach_copy_ lowering


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "all_gather_bucketing fx pass"



Porting passes to bucket all_gathers

The main logic of the pass is done via 
1. Searching for all all_gathers from the buckets


Copying tests from wconstab PR to test compatibility with reordering.
Test checks only compatibility, as because of (3) the joint all_gather will be scheduled already as early as possible and no space for reordering.

Pass changes:
Using mutation ops to match performance of fsdp, in future the perfect scenario will be to have only functional graph, that inductor does all memory optimizations on its own without mutable ops.


Inductor changes:
Adding foreach_copy_ lowering


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

IvanKobzarev committed:
Update on "all_gather_bucketing fx pass"



Porting passes to bucket all_gathers

The main logic of the pass is done via 
1. Searching for all all_gathers from the buckets


Copying tests from wconstab PR to test compatibility with reordering.
Test checks only compatibility, as because of (3) the joint all_gather will be scheduled already as early as possible and no space for reordering.

Pass changes:
Using mutation ops to match performance of fsdp, in future the perfect scenario will be to have only functional graph, that inductor does all memory optimizations on its own without mutable ops.


Inductor changes:
Adding foreach_copy_ lowering


cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]
